src/irc/core/sasl.c

```diff
@@ -105,49 +113,149 @@ static void sasl_success(IRC_SERVER_REC *server, const char *data, const char *f
 	cap_finish_negotiation(server);
 }
 
-static void sasl_step(IRC_SERVER_REC *server, const char *data, const char *from)
+/*
+ * Responsible for reassembling incoming SASL requests. SASL requests must be split
+ * into 400 byte requests to stay below the IRC command length limit of 512 bytes.
+ * The spec says that if there is 400 bytes, then there is expected to be a
+ * continuation in the next chunk. If a message is exactly a multiple of 400 bytes,
+ * there must be a blank message of "AUTHENTICATE +" to indicate the end.
+ *
+ * This function returns the fully reassembled and decoded AUTHENTICATION message if
+ * completed or NULL if there are more messages expected.
+ */
+static GBytes *sasl_reassemble_incoming(IRC_SERVER_REC *server, const char *enc_req)
 {
-	IRC_SERVER_CONNECT_REC *conn;
-	GString *req;
-	char *enc_req;
+	char *buffer;
+	GBytes *decoded;
+
+	/* Check if there is an existing fragment to prepend. */
+	buffer = g_hash_table_lookup(sasl_buffers, server->tag);
+	if (buffer != NULL) {
+		if (g_strcmp0("+", enc_req) == 0) {
+			enc_req = buffer;
+		} else {
+			enc_req = g_strconcat(buffer, enc_req, NULL);
```

Comment:
This replaces the received end_req with a longer version, so the check for `strlen(enc_req) == AUTHENTICATE_CHUNK_SIZE` below will never pass after the first chunk.

It also sets a string that should be freed (and is leaking) in a variable that shouldn't.